src/control/lib/spdk/ctests/nvme_control_ut.c

```diff
@@ -0,0 +1,286 @@
+test_nvme_discover_set_g_controllers(void **state)
+{
+	struct ret_t *ret;
+
+	(void) state; /*unused*/
+
+	assert_null(g_controllers);
+	g_controllers = malloc(sizeof(struct ctrlr_entry));
+	g_controllers->ctrlr = NULL;
+	g_controllers->dev_health = NULL;
+	g_controllers->next = NULL;
+
+	ret = _nvme_discover(&mock_spdk_nvme_probe_ok,
+			     &mock_spdk_nvme_detach_ok,
+			     &mock_get_dev_health_logs);
+	assert_int_equal(ret->rc, 0);
+
+	assert_null(ret->ctrlrs);
+
+	free(ret);
+}
+
+static void
+test_nvme_discover_probe_fail(void **state)
+{
+	struct ret_t *ret;
+
+	(void) state; /*unused*/
+
+	assert_null(g_controllers);
+	g_controllers = malloc(sizeof(struct ctrlr_entry));
+	g_controllers->ctrlr = NULL;
+	g_controllers->dev_health = NULL;
+	g_controllers->next = NULL;
+
+	ret = _nvme_discover(&mock_spdk_nvme_probe_fail,
+			     &mock_spdk_nvme_detach_ok,
+			     &mock_get_dev_health_logs);
+	assert_int_equal(ret->rc, -1);
+
+	assert_null(ret->ctrlrs);
+
+	free(ret);
+}
+
+static void
+test_nvme_collect(void **state)
+{
+	struct ret_t		       *ret;
+	struct spdk_nvme_ctrlr_opts	opts = {};
+	struct spdk_nvme_transport_id	trid1 = {};
+	struct spdk_nvme_transport_id	trid2 = {};
+	struct spdk_nvme_ctrlr		ctrlr1;
+	struct spdk_nvme_ctrlr		ctrlr2;
+	struct ctrlr_entry	       *entry1;
+	struct ctrlr_entry	       *entry2;
+
+	(void) state;
+
+	ret = init_ret();
+
+	memset(&ctrlr1, 0, sizeof(struct spdk_nvme_ctrlr));
+	snprintf(ctrlr1.trid.traddr, sizeof(ctrlr1.trid.traddr),
+		 "0000:01:00.0");
+	ctrlr1.trid.trtype = SPDK_NVME_TRANSPORT_PCIE;
+	memset(&ctrlr2, 0, sizeof(struct spdk_nvme_ctrlr));
+	snprintf(ctrlr2.trid.traddr, sizeof(ctrlr2.trid.traddr),
+		 "0000:02:00.0");
+	ctrlr2.trid.trtype = SPDK_NVME_TRANSPORT_PCIE;
+
+	memset(&trid1, 0, sizeof(trid1));
+	trid1.trtype = SPDK_NVME_TRANSPORT_PCIE;
+	snprintf(trid1.traddr, sizeof(trid1.traddr),
+		 "0000:01:00.0");
+	memset(&trid2, 0, sizeof(trid2));
+	trid2.trtype = SPDK_NVME_TRANSPORT_PCIE;
+	snprintf(trid2.traddr, sizeof(trid2.traddr),
+		 "0000:02:00.0");
+
+	/* attach_cb gets called with matching trid & ctrlr.trid by probe for
+	 * each controller on the bus
+	 */
+	assert_null(g_controllers);
+	attach_cb(NULL, &trid1, &ctrlr1, &opts);
+	assert_non_null(g_controllers);
+	entry1 = g_controllers;
+	attach_cb(NULL, &trid2, &ctrlr2, &opts);
+	assert_non_null(g_controllers);
+	entry2 = g_controllers;
+	assert_ptr_not_equal(entry1, entry2);
+
+	assert_null(ret->ctrlrs);
+	_collect(ret, &mock_spdk_nvme_ctrlr_get_data,
+		 &mock_spdk_nvme_ctrlr_get_pci_device,
+		 &mock_spdk_pci_device_get_socket_id);
+
+	if (ret->rc != 0)
+		fprintf(stderr, "collect err: %s\n", ret->err);
+	assert_return_code(ret->rc, 0);
+	assert_non_null(ret->ctrlrs);
+	assert_string_equal(ret->ctrlrs->pci_addr, "0000:01:00.0");
+	assert_non_null(ret->ctrlrs->next);
+	assert_string_equal(ret->ctrlrs->next->pci_addr, "0000:02:00.0");
+	assert_null(ret->ctrlrs->next->next);
+
+	free(ret);
+}
+
+int
+teardown(void ** state)
```

Comment:
(style)  "foo ** bar" should be "foo **bar"

Reply:
Just remove space after **